core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -410,6 +409,7 @@ public void commit() {
                   // to ensure that if a concurrent operation assigns the UUID, this operation will
                   // not fail.
                   taskOps.commit(base, updated.withUUID());
+                  committedSnapshot.set(newSnapshot);
```

Comment:
I think the code was written using `newSnapshotId` because we want to strictly limit what happens after `commit` is called. After `commit` succeeds, the table state has changed and we want to reduce the risk that anything happens that would cause the client to mistakenly think that the commit has failed. That's why `commit` is the last thing called in this block.

It's not incorrect to change it to what you have, but I definitely prefer the original way.

Reply:
Without doing this, the downside is we have to re-read from objectstorage the snapshot we just wrote on every commit, which also is not good.

Writing to an atomic reference is lightweight. Feels like a reasonable tradeoff. WDYT?

Reply:
The big problem is that this is set after `commit` is called. Can you refactor to avoid that? You'll just have to make sure that the commit succeeded in all cases before using the snapshot since you no longer have the guarantee that the snapshot was found in the table metadata.

Reply:
The only place the snapshot is used is after the commit succeeds, so was an easy fix. PTAL